Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py

```diff
@@ -0,0 +1,251 @@
+                )
+            for remote, branches in cls.branches_on(repository, commit.hash).items():
+                if remote is not None and remote not in repository.source_remotes():
+                    continue
+                for branch in branches:
+                    commit = repository.find('remotes/{}/{}'.format(remote, branch) if remote else branch)
+                    cls._push_branch_ref(
+                        mapping=branches_to_publish,
+                        repository=repository,
+                        branch=branch, commit=commit,
+                    )
+
+        if tags_to_publish - requested_refs:
+            log.warning("Inferred the following tags:")
+            for tag in tags_to_publish - requested_refs:
+                log.warning('    {}'.format(tag))
+            log.warning('')
+
+        if set(branches_to_publish.keys()) - requested_refs:
+            log.warning("Inferred the following branches:")
+            for branch in set(branches_to_publish.keys()) - requested_refs:
+                log.warning('    {} @ {}'.format(branch, branches_to_publish[branch]))
+            log.warning('')
+
+        if tags_to_publish:
+            print("Pushing the following tags to {}:".format(args.remote))
+            for tag in sorted(tags_to_publish):
+                print('    {}'.format(tag))
+            print('')
+        if branches_to_publish:
+            print("Pushing the branches at the following commits to {}:".format(args.remote))
+            for branch in sorted(branches_to_publish.keys()):
+                print('    {} @ {}'.format(branch, branches_to_publish[branch]))
+            print('')
+
+        if Terminal.choose(
+            "Are you sure you would like to publish to '{}' remote?".format(args.remote),
+            default='No',
+        ) == 'No':
+            sys.stderr.write("Publication canceled\n")
+            return 1
+
+        return_code = 0
+        for branch, commit in branches_to_publish.items():
+            print('pushing {} to {}...'.format(branch, args.remote))
+            if run(
+                [repository.executable(), 'push', args.remote, '{}:refs/heads/{}'.format(commit.hash, branch)],
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            ).returncode:
+                sys.stderr.write('Failed to push {} to {}\n'.format(tag, args.remote))
+                return_code += 1
+        for tag in tags_to_publish:
+            print('pushing {} to {}...'.format(tag, args.remote))
+            if run(
+                [repository.executable(), 'push', args.remote, tag],
+                cwd=repository.root_path,
+                capture_output=True,
+                encoding='utf-8',
+            ).returncode:
+                sys.stderr.write('Failed to push {} to {}\n'.format(tag, args.remote))
+                return_code += 1
```

Comment:
Why not push these all in one git invocation, with --atomic, so that the remote isn't left in a indeterminate state if we are interrupted?

Reply:
Good idea, always forget that `--atomic` is a thing....